packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,486 @@
+    DidExchangeStateMachine.assertCreateMessageState(DidExchangeCompleteMessage.type, connectionRecord)
+
+    const threadId = connectionRecord.threadId
+    const parentThreadId = connectionRecord.invitation?.id
+
+    if (!threadId) {
+      throw new AriesFrameworkError(`Connection record ${connectionRecord.id} does not have 'threadId' attribute.`)
+    }
+
+    if (!parentThreadId) {
+      throw new AriesFrameworkError(
+        `Connection record ${connectionRecord.id} does not have 'parentThreadId' attribute.`
+      )
+    }
+
+    const message = new DidExchangeCompleteMessage({ threadId, parentThreadId })
+
+    await this.updateState(DidExchangeCompleteMessage.type, connectionRecord)
+    this.logger.debug(`Create message ${DidExchangeCompleteMessage.type} end`, { connectionRecord, message })
+    return message
+  }
+
+  public async processComplete(
+    messageContext: InboundMessageContext<DidExchangeCompleteMessage>
+  ): Promise<ConnectionRecord> {
+    this.logger.debug(`Process message ${DidExchangeCompleteMessage.type} start`, messageContext)
+    const { connection: connectionRecord, message } = messageContext
+
+    if (!connectionRecord) {
+      throw new AriesFrameworkError('No connection record in message context.')
+    }
+
+    DidExchangeStateMachine.assertProcessMessageState(DidExchangeCompleteMessage.type, connectionRecord)
+
+    if (connectionRecord.invitation?.id !== message.thread?.parentThreadId) {
+      throw new DidExchangeProblemReportError('Missing reference to invitation.', {
+        problemCode: DidExchangeProblemReportReason.CompleteRejected,
+      })
+    }
+
+    await this.updateState(DidExchangeCompleteMessage.type, connectionRecord)
+    this.logger.debug(`Process message ${DidExchangeCompleteMessage.type} end`, { connectionRecord })
+    return connectionRecord
+  }
+
+  private async updateState(messageType: string, connectionRecord: ConnectionRecord) {
+    this.logger.debug(`Updating state`, { connectionRecord })
+    const nextState = DidExchangeStateMachine.nextState(messageType, connectionRecord)
+    return this.connectionService.updateState(connectionRecord, nextState)
+  }
+
+  private async createPeerDidDoc(routing: Routing) {
+    const publicKeyBase58 = routing.verkey
+
+    const ed25519Key = Key.fromPublicKeyBase58(publicKeyBase58, KeyType.Ed25519)
+    const x25519Key = Key.fromPublicKey(convertPublicKeyToX25519(ed25519Key.publicKey), KeyType.X25519)
+
+    // For peer dids generated with method 1, the controller MUST be #id as we don't know the did yet
+    const ed25519VerificationMethod = getEd25519VerificationMethod({
+      id: uuid(),
+      key: ed25519Key,
+      controller: '#id',
+    })
+    const x25519VerificationMethod = getX25519VerificationMethod({
+      id: uuid(),
+      key: x25519Key,
+      controller: '#id',
+    })
+
+    let mediatorRoutingKey
+    if (routing.routingKeys.length > 0) {
+      const [mediatorPublicKeyBase58] = routing.routingKeys
```

Comment:
As I was working on iteration over endpoints as mentioned below I found out that this is a similar problem. I should perhaps transform all routing keys in this way, right? I'm just not sure why we're using did:key for routing keys and not for recipient keys. 🤔

Reply:
Hmm you're right that doesn't make sense. From looking at the did:key RFC we shouldn't use did:key encoding at all when using did:peer did docs, but rather use a key reference. However what should we put there in that case? We don't have a key reference for the mediator routing keys.

I'm confused with the mediator coordination protocol, which requires the routing keys to be encoded as did:keys: https://github.com/hyperledger/aries-rfcs/blob/main/features/0211-route-coordination/README.md#mediation-grant.

However the [did:key RFC](https://github.com/hyperledger/aries-rfcs/blob/main/features/0360-use-did-key/README.md) mentions:

> Service entries in did:peer DIDDocs (such as in RFCs [0094-cross-domain-messaging](https://github.com/hyperledger/aries-rfcs/tree/main/concepts/0094-cross-domain-messaging) and [0067-didcomm-diddoc-conventions](https://github.com/hyperledger/aries-rfcs/tree/main/features/0067-didcomm-diddoc-conventions)) should NOT use a did:key public key representation. Instead, service entries in the DIDDoc should reference keys defined internally in the DIDDoc where appropriate.

It's not always possible to include key references from did docs, as that would mean the mediator needs to use a public did. So in that case I'm guessing we should use a did:key reference so we can still make a resolvable key from the routing keys. If we directly use the publicKeyBase58 there's no way to determine which key type is used. 

See also this issue for a relevant discussion: https://github.com/hyperledger/aries-rfcs/issues/704

Basically boils down to:
- DIDComm v2 will support any type of did for the routingKeys, allowing mediators to rotate keys etc.. without needing to communicate with the recipient
- DIDComm v2 doesn't have the recipientKeys anymore, so it's mostly a leftover legacy thing. Also having recipientKeys outside of the did document makes less sense than having routingKeys outside of the did document 


The encoding of the `recipientKeys` should be key references.  So instead of this:
```
recipientKeys: [ed25519Key.publicKeyBase58],
```

it should be this:

```
recipientKeys: [`#${ed25519VerificationMethod.id}`],
```

Well actually, this:

```
recipientKeys: [`#${x25519VerificationMethod.id}`],
```

but that will cause other problems

Reply:
What other problems do you have on your mind? I guess one is that we cannot convert x25519 to ed25519 needed by Indy?

It seems that ACA-Py supports just [ed25519 base58](https://github.com/hyperledger/aries-cloudagent-python/blob/main/aries_cloudagent/connections/base_manager.py#L139).

I aligned the recipient and routing keys, it's using `ed25519Key.publicKeyBase58` in service for both. I also added the routing key into the did doc as a verification method.

This is tested with added oob mediation test. I think it's a good starting point before we finalize how it should work according to the current and future crypto libs and protocols.